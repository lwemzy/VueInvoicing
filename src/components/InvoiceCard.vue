<template>
  <v-form ref="form" v-model="validForm" lazy-validation>
    <v-card class="pa-2">
      <v-card-title>
        <span class="headline">{{ formTitle }}</span>
      </v-card-title>
      <v-card-text>
        <v-container>
          <v-row>
            <v-col cols="12">
              <v-text-field
                :rules="invoiceRules"
                v-model="editedInvoice.invoice_no"
                label="Invoice Number"
                required
              ></v-text-field>
            </v-col>
            <v-col cols="12">
              <v-menu
                ref="menu"
                v-model="menu"
                :close-on-content-click="false"
                :return-value.sync="editedInvoice.due_date"
                transition="scale-transition"
                offset-y
                min-width="290px"
              >
                <template v-slot:activator="{ on }">
                  <v-text-field
                    v-model="editedInvoice.due_date"
                    label="Due Date"
                    prepend-icon="mdi-calendar"
                    readonly
                    v-on="on"
                  ></v-text-field>
                </template>
                <v-date-picker v-model="editedInvoice.due_date" no-title scrollable>
                  <div class="flex-grow-1"></div>
                  <v-btn text color="primary" @click="menu = false">Cancel</v-btn>
                  <v-btn text color="primary" @click="$refs.menu.save(editedInvoice.due_date)">OK</v-btn>
                </v-date-picker>
              </v-menu>
            </v-col>
          </v-row>
        </v-container>
      </v-card-text>
      <v-card-actions>
        <div class="flex-grow-1"></div>
        <v-btn color="blue darken-1" text @click="close">Cancel</v-btn>
        <v-btn color="blue darken-1" text @click="save">Save</v-btn>
      </v-card-actions>
    </v-card>
  </v-form>
</template>

<script>
export default {
  name: "InvoiceCard",
  props: ["editedInvoice", "formTitle", "menu", "invoiceRules", "valid"],
  data() {
    return {
      validForm: this.valid
    };
  },
  methods: {
    close() {
      this.$emit("close");
    },
    save() {
      if (this.$refs.form.validate()) {
        this.$emit("save");
      }
    }
  }
};
</script> 