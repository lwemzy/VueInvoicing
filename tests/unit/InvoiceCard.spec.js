// Libraries
import Vue from 'vue';
import Vuetify from 'vuetify';
Vue.use(Vuetify);
// Components
import InvoiceCard from '@/components/InvoiceCard';

// Utilities
import { mount, createLocalVue } from '@vue/test-utils';

const localVue = createLocalVue();

describe('InvoiceCard.vue', () => {
  let vuetify;

  beforeEach(() => {
    vuetify = new Vuetify();
  });

  it('renders a vue instance', () => {
    const wrapper = mount(InvoiceCard, {
      localVue,
      vuetify,
      propsData: {
        formTitle: 'Foobar',
        editedInvoice: {
          invoice_no: 1213,
          due_date: new Date().toISOString().substr(0, 10)
        }
      }
    });
    expect(wrapper.isVueInstance()).toBe(true);
  });

  it('should have a custom title and match snapshot', () => {
    const wrapper = mount(InvoiceCard, {
      localVue,
      vuetify,
      propsData: {
        formTitle: 'Foobar',
        editedInvoice: {
          invoice_no: 1213,
          due_date: new Date().toISOString().substr(0, 10)
        }
      }
    });

    expect(wrapper.html()).toMatchSnapshot();

    const title = wrapper.find('.v-card__title > span');

    expect(title.text()).toBe('Foobar');
  });

  it('v-btn is clickable', () => {
    const wrapper = mount(InvoiceCard, {
      localVue,
      vuetify,
      propsData: {
        formTitle: 'Foobar',
        editedInvoice: {
          invoice_no: 1213,
          due_date: new Date().toISOString().substr(0, 10)
        }
      }
    });

    const button = wrapper.find('.v-btn');

    button.trigger('click');
  });

  it('Input Accepts User input', () => {
    const wrapper = mount(InvoiceCard, {
      localVue,
      vuetify,
      propsData: {
        formTitle: 'Foobar',
        editedInvoice: {
          invoice_no: 1213,
          due_date: new Date().toISOString().substr(0, 10)
        }
      }
    });

    const newValue = 'Updated value';
    const input = wrapper.find('input');
    input.element.value = newValue;
    input.trigger('change');
  });
});
