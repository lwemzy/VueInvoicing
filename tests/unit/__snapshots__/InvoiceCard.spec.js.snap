// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvoiceCard.vue should have a custom title and match snapshot 1`] = `
<div class="pa-2 v-card v-sheet theme--light">
  <div class="v-card__title"><span class="headline">Foobar</span></div>
  <div class="v-card__text">
    <div class="container">
      <div class="row">
        <div class="col col-12">
          <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field">
            <div class="v-input__control">
              <div class="v-input__slot">
                <div class="v-text-field__slot"><label for="input-20" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Invoice Number</label><input required="required" id="input-20" type="text"></div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
              </div>
            </div>
          </div>
        </div>
        <div class="col col-12">
          <div class="v-menu">
            <div class="v-input v-input--is-label-active v-input--is-dirty v-input--is-readonly theme--light v-text-field">
              <div class="v-input__prepend-outer">
                <div class="v-input__icon v-input__icon--prepend"><i aria-hidden="true" class="v-icon notranslate mdi mdi-calendar theme--light"></i></div>
              </div>
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-text-field__slot"><label for="input-24" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Due Date</label><input id="input-24" readonly="readonly" type="text"></div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
                </div>
              </div>
            </div>
            <div role="menu" class="v-menu__content theme--light " style="max-height: auto; min-width: 290px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"></div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="v-card__actions">
    <div class="flex-grow-1"></div> <button type="button" class="v-btn v-btn--flat v-btn--text theme--light v-size--default blue--text text--darken-1"><span class="v-btn__content">Cancel</span></button> <button type="button" class="v-btn v-btn--flat v-btn--text theme--light v-size--default blue--text text--darken-1"><span class="v-btn__content">Save</span></button>
  </div>
</div>
`;
